README.md
# kindui

---

A nice CMD module.

---

## Usage

It is very easy to use this module.

````html
<div class="kind-kindui">
</div>
````

```css
.kindui{
    color:red;
}

```

## Api

Here is some details.


HISTORY.md
# History

---

## 1.0.0

`new` kind/kindui First version.


test/js/BoxModel.js
/*
 * @author: human.huang
 * @created date: 2015-04-07
 */

var BoxModel = (function () {

      var boxCount = 1;

      Pix.default = {

      };

      function Pix(dom)  {

            if (dom === undefined) {
                  throw new Error('Must have dom arguments.');
            }
            else if (typeof dom == 'string') {
                  dom = document.querySelector(dom);
                  if(!dom) {
                        throw new Error('Invalid dom selector');
                  }
            }
            else if(typeof dom == 'object' && dom.length) {
                  dom.forEach((function(oneDom) {
                        new Pix(oneDom);
                  }));
                  return ;
            }


            if (this.constructor != arguments.callee) {
                  return new Pix(arguments[0])
            }
            this.dom = dom;
            this.init();
            return this;
      };

      Pix.prototype.init = function () {

            this.currentTipBox = null;
            this.currentTipBoxClass = null;

            this._injectCommonStyle();

            this._initBoxModel();

            this._getDomOffset();

            return this;

      };
      Pix.prototype.showBox = function() {
            this.createMtTip();
            this.createPtTip();

            this.createMlTip();
            this.createPlTip();
      },
      Pix.prototype._injectCommonStyle = function () {

            var style = /\/\*([\s\S]*)\*\//mg.exec(this._commonStyle.toString())[1];
            this._addStyleSheet(style);
      };

      Pix.prototype._appendStyleToDom = function (dom, attrs) {
            attrs.forEach(function (item, i) {
                  item[0] = item[0].replace(/-(\w)/g, function (_, letter) {
                        return letter.toUpperCase()
                  });
                  dom.style[item[0]] = item[1];
            });
      };

      Pix.prototype._generateStyle = function (className, cssAttrs) {

            var cssStr = '.' + className + '{';
            cssAttrs.forEach(function (item) {
                  cssStr += item[0] + ':' + item[1] + ';';
            });
            cssStr += '}';

            this._addStyleSheet(cssStr);
      };
      Pix.prototype._commonStyle = function () {
            /*

             .kind-tip-box{
                   position: absolute;
             }

             .kind-tip-box-arrow-up{
                   position: absolute;
             }

             .kind-tip-box-arrow-line{
                   position: absolute;
                   background: red;
             }

             .kind-tip-box-arrow-down{
                   position: absolute;
             }

             .kind-tip-box-indicator-line{
                   display: block;
                   position: absolute;
                   background: red;
                   transform-origin: 0 0;
             }

             .kind-tip-box-text{
                   display: block;
                   font-size: 12px;
                   height: 12px;
                   position: absolute;
                   background: red;
                   color: #ffffff;
                   white-space: nowrap;
             }

             .kind-tip-box-indicator-line-arrow{
                   position: absolute;
             }

             */
      };
      Pix.prototype._boxStyleUpdown = function () {
            /*

             .{classPrefix}{
                   left: {boxLeft}px;
                   top: {boxTop}px;
                   height: {boxHeight}px;
                   width: {boxWidth}px;
                   border-top: 1px solid red;
                   border-bottom: 1px solid red;
             }
             .{classPrefix} .kind-tip-box-arrow-up{
                   top: {arrowUpTop}px;
                   left: {arrowUpLeft}px;
                   border-bottom: 4px solid red;
                   border-left: 4px solid transparent;
                   border-right: 4px solid transparent;
             }

             .{classPrefix} .kind-tip-box-arrow-line{
                   top: {arrowLineTop}px;
                   left: {arrowLineLeft}px;
                   height: {arrowLineHeight}px;
                   width: {arrowLineWidth}px;
             }

             .{classPrefix} .kind-tip-box-arrow-down{
                   bottom: {arrowDownBottom}px;
                   left: {arrowDownLeft}px;
                   border-top: 4px solid red;
                   border-left: 4px solid transparent;
                   border-right: 4px solid transparent;
             }

             .{classPrefix} .kind-tip-box-indicator-line{
                   left: {indicatorLineLeft}px;
                   top: {indicatorLineTop}px;
                   width: {indicatorLineWidth}px;
                   height: {indicatorLineHeight}px;
                   transform: {indicatorLineTransform};
             }

             .{classPrefix} .kind-tip-box-text{
                   left: {textLeft}px;
                   top: {textTop}px;
             }

             .{classPrefix} .kind-tip-box-indicator-line-arrow{
                   left: {indicatorLineArrowLeft}px;
                   top: {indicatorLineArrowTop}px;
                   border-left: 4px solid red;
                   border-top: 4px solid transparent;
                   border-bottom: 4px solid transparent;
                   transform-origin: 50% 0;
                   transform:{indicatorArrowTransform}
             }

             */
      };
      Pix.prototype._boxStyleLeftRight = function () {
            /*

             .{classPrefix}{
                   left: {boxLeft}px;
                   top: {boxTop}px;
                   height: {boxHeight}px;
                   width: {boxWidth}px;
                   border-left: 1px solid red;
                   border-right: 1px solid red;
             }
             .{classPrefix} .kind-tip-box-arrow-up{
                   top: {arrowUpTop}px;
                   left: {arrowUpLeft}px;

                   border-bottom: 4px solid transparent;
                   border-top: 4px solid transparent;
                   border-right: 4px solid red;
             }

             .{classPrefix} .kind-tip-box-arrow-line{
                   top: {arrowLineTop}px;
                   left: {arrowLineLeft}px;
                   height: {arrowLineHeight}px;
                   width: {arrowLineWidth}px;
             }

             .{classPrefix} .kind-tip-box-arrow-down{
                   left:auto;
                   top: {arrowDownTop}px;
                   right: {arrowDownRight}px;
                   border-bottom: 4px solid transparent;
                   border-top: 4px solid transparent;
                   border-left: 4px solid red;
             }

             .{classPrefix} .kind-tip-box-indicator-line{
             left: {indicatorLineLeft}px;
             top: {indicatorLineTop}px;
             width: {indicatorLineWidth}px;
             height: {indicatorLineHeight}px;
             transform: {indicatorLineTransform};
             }

             .{classPrefix} .kind-tip-box-text{
             left: {textLeft}px;
             top: {textTop}px;
             }

             .{classPrefix} .kind-tip-box-indicator-line-arrow{
             left: {indicatorLineArrowLeft}px;
             top: {indicatorLineArrowTop}px;
             border-top: 4px solid red;
             border-left: 4px solid transparent;
             border-right: 4px solid transparent;
             transform-origin: 50% 0;
             transform:{indicatorArrowTransform}
             }

             */
      }

      Pix.prototype._createTipBox = function (text) {

            var textPrefix = arguments.callee.caller.toString().match(/this\._createTipBox\(this\.(\w{2})/)[1];

            var CONF = {
                  mt:'margin-top',
                  ml:'margin-left',
                  pt:'padding-top',
                  pl:'padding-left'
            };
            textPrefix = CONF[textPrefix];

            var boxCountClass = this.currentTipBoxClass = 'kind-tip-box-' + boxCount++;
            var tipBox = this.currentTipBox = document.createElement('div');
            tipBox.className = 'kind-tip-box ' + boxCountClass;
            tipBox.innerHTML = [
                  '<div class="kind-tip-box-arrow-up"></div>',
                  '<div class="kind-tip-box-arrow-line"></div>',
                  '<div class="kind-tip-box-arrow-down"></div>',
                  '<div class="kind-tip-box-indicator-line"></div>',
                  '<div class="kind-tip-box-indicator-line-arrow"></div>',
                  '<div class="kind-tip-box-text">'+this.dom.nodeName + '|' +textPrefix +':'+text+'</div>'
            ].join('');

            document.body.appendChild(tipBox);
      };
      Pix.prototype.createMtTip = function () {

            if (this.mt == 0) {
                  return;
            }

            this._createTipBox(this.mt + 'px');

            var attrs = {
                  classPrefix:this.currentTipBoxClass,

                  boxLeft:this.left,
                  boxTop:this.top - this.mt,
                  boxWidth:10,
                  boxHeight:this.mt,

                  arrowUpTop:0,
                  arrowUpLeft:10 - 5 - 4,

                  arrowDownBottom:0,
                  arrowDownLeft:10 - 5 - 4,

                  arrowLineTop:0,
                  arrowLineLeft:5 ,
                  arrowLineHeight:this.mt,
                  arrowLineWidth:1,


                  indicatorLineLeft:10 /2,
                  indicatorLineTop:this.mt / 2,
                  indicatorLineWidth:40,
                  indicatorLineHeight:1,
                  indicatorLineTransform:'rotateZ(-30deg)',

                  indicatorLineArrowLeft:10 / 2 + Math.sqrt(3) / 2 * 40 - 4,
                  indicatorLineArrowTop:(this.mt / 2 - 20 - 4) ,
                  indicatorArrowTransform:'rotateZ(-30deg)',

                  textLeft:10 / 2 + Math.sqrt(3) / 2 * 40,
                  textTop:-(40 / 2 - this.mt / 2 + 12)

            };


            var boxStyle = /\/\*([\s\S]*)\*\//mg.exec(this._boxStyleUpdown.toString())[1];

            var boxStyled = boxStyle.replace(/{(\w+)}/gm, function(text, rText) {
                  return attrs[rText];
            });

            this._addStyleSheet(boxStyled);

      }

      Pix.prototype.createMlTip = function ()   {

            if (this.ml == 0) {
                  return;
            }

            this._createTipBox(this.ml + 'px');

            var attrs = {
                  classPrefix:this.currentTipBoxClass,

                  boxLeft:this.left - this.ml,
                  boxTop:this.top + this.dom.clientHeight - 10,
                  boxWidth:this.ml,
                  boxHeight:10,

                  arrowUpTop:5-4,
                  arrowUpLeft:0,

                  arrowDownTop:5-4,
                  arrowDownRight: 0,

                  arrowLineTop:5,
                  arrowLineLeft:0,
                  arrowLineHeight:1,
                  arrowLineWidth:this.ml,


                  indicatorLineLeft:this.ml /2,
                  indicatorLineTop:5,
                  indicatorLineWidth:1,
                  indicatorLineHeight:20,
                  indicatorLineTransform:'',

                  indicatorLineArrowLeft:this.ml/2 -4 ,
                  indicatorLineArrowTop: 20 + 4,
                  indicatorArrowTransform:'',

                  textLeft:this.mt / 2  - this.currentTipBox.querySelector('.kind-tip-box-text').clientWidth / 2,
                  textTop:20 + 5 + 2,

            };


            var boxStyle = /\/\*([\s\S]*)\*\//mg.exec(this._boxStyleLeftRight.toString())[1];

            var boxStyled = boxStyle.replace(/{(\w+)}/gm, function(text, rText) {
                  return attrs[rText];
            });

            this._addStyleSheet(boxStyled);

      }

      Pix.prototype.createPlTip = function () {

            if (this.pl == 0) {
                  return;
            }

            this._createTipBox(this.pl + 'px');

            var attrs = {
                  classPrefix:this.currentTipBoxClass,

                  boxLeft:this.left,
                  boxTop:this.top + this.dom.clientHeight - 10,
                  boxWidth:this.pl,
                  boxHeight:10,

                  arrowUpTop:5-4,
                  arrowUpLeft:0,

                  arrowDownTop:5-4,
                  arrowDownRight: 0,

                  arrowLineTop:5,
                  arrowLineLeft:0,
                  arrowLineHeight:1,
                  arrowLineWidth:this.pl,


                  indicatorLineLeft:this.pl /2,
                  indicatorLineTop:5,
                  indicatorLineWidth:40,
                  indicatorLineHeight:1,
                  indicatorLineTransform:'rotateZ(30deg)',

                  indicatorLineArrowLeft:this.pl / 2 + Math.sqrt(3) / 2 * 40 - 4,
                  indicatorLineArrowTop: 20 + 5,
                  indicatorArrowTransform:'rotateZ(-30deg)',

                  textLeft:this.pl / 2 + Math.sqrt(3) / 2 * 40 + 2,
                  textTop:20 + 5 + 2,

            };


            var boxStyle = /\/\*([\s\S]*)\*\//mg.exec(this._boxStyleLeftRight.toString())[1];

            var boxStyled = boxStyle.replace(/{(\w+)}/gm, function(text, rText) {
                  return attrs[rText];
            });

            this._addStyleSheet(boxStyled);

      }
      Pix.prototype.createPtTip = function () {

            if (this.pt == 0) {
                  return;
            }

            this._createTipBox(this.pt + 'px');

            var attrs = {
                  classPrefix:this.currentTipBoxClass,

                  boxLeft:this.left,
                  boxTop:this.top,
                  boxWidth:10,
                  boxHeight:this.pt,

                  arrowUpTop:0,
                  arrowUpLeft:10 - 5 - 4,

                  arrowDownBottom:0,
                  arrowDownLeft:10 - 5 - 4,

                  arrowLineTop:0,
                  arrowLineLeft:5 ,
                  arrowLineHeight:this.pt,
                  arrowLineWidth:1,


                  indicatorLineLeft:10 /2,
                  indicatorLineTop:this.pt /2,
                  indicatorLineWidth:40,
                  indicatorLineHeight:1,
                  indicatorLineTransform:'',

                  indicatorLineArrowLeft:40+4,
                  indicatorLineArrowTop:this.pt / 2 - 4,
                  indicatorArrowTransform:'',

                  textLeft:40+4+4,
                  textTop:this.pt / 2 - 6,

            };


            var boxStyle = /\/\*([\s\S]*)\*\//mg.exec(this._boxStyleUpdown.toString())[1];

            var boxStyled = boxStyle.replace(/{(\w+)}/gm, function(text, rText) {
                  return attrs[rText];
            });

            this._addStyleSheet(boxStyled);

      };

      Pix.prototype._addStyleSheet = function (styleStyle) {
            var style = document.createElement('style');
            style.innerHTML = styleStyle;
            //IE hack cssText;
            document.head.appendChild(style);
      };

      Pix.prototype._initBoxModel = function () {
            var self = this;
            var boxes = [
                  ['mt', 'marginTop'],
                  ['mb', 'marginBottom'],
                  ['ml', 'marginLeft'],
                  ['mr', 'marginRight'],

                  ['pt', 'paddingTop'],
                  ['pb', 'paddingBottom'],
                  ['pl', 'paddingLeft'],
                  ['pr', 'paddingRight']

            ];

            boxes.forEach(function (item, i) {
                  self[item[0]] = self._getStyle(item[1]);
            });

      };
      Pix.prototype._getDomOffset = function () {
            var rect = this.dom.getBoundingClientRect(),
                  docOffset = this._getDocOffset();
            this.top = rect.top + docOffset.top,
                  this.left = rect.left + docOffset.left;

      };
      Pix.prototype._getDocOffset = function () {
            return {
                  left: document.documentElement.scrollLeft || document.body.scrollLeft,
                  top: document.documentElement.scrollTop || document.body.scrollTop
            }
      };
      Pix.prototype._getStyle = function (styleName) {
            var styleValue = getComputedStyle(this.dom, null)[styleName];
            if (styleValue == null) {
                  return null;
            }
            return parseFloat(styleValue);
      };

      return Pix;
})();


test/ajax_encoding/get_utf8.php
{"a":"啊"}

page/js/doc.js
function convert2Hex(rgb) {
      if (~rgb.indexOf('#')) {
            return rgb;
      }
      var rgbArr = rgb.match(/\d+/g);
      return '#' + convert(rgbArr[0]) + convert(rgbArr[1]) + convert(rgbArr[2]);

      function convert(d) {
            var hex = parseInt(d).toString(16);
            if (hex.length == 1) {
                  return '0' + hex;
            }
            return hex;
      }
}

Array.prototype.unique = function()
{
      var n = [];
      for(var i = 0; i < this.length; i++)
      {
            if (n.indexOf(this[i]) == -1) n.push(this[i]);
      }
      return n;
}

function parseTable(conf) {

      var head = renderTableHead(conf.headCols), subTitle = '';


      if (conf.title) {
            if (conf.subTitle) {
                  subTitle = '<span class="ft-gray ft-14 ml10">' + conf.subTitle + '</span>'
            }
            head = '<h3 class="table-title">' + conf.title + subTitle + '</h3>' + head;
      }

      var body = renderTableBody(conf.body);

      var table = renderTable(head, body);

      var $table = $(table);

      if (conf.onRendered) {
            conf.onRendered($table);
      }

      return $(table);
      function renderTableHead(arr) {
            var head = '<thead><tr>';

            $.each(arr, function (_, item) {
                  head += '<th>' + item.text + '</th>';
            });
            head += '</tr></thead>';

            return head;
      }


      function renderTableBody(arr) {
            var body = '<tbody>';

            $.each(arr, function (_, itemTR) {

                  var tr = '<tr>';
                  $.each(itemTR, function (_, itemTD) {

                        var classFrag = '', styleFrag = '', attrsFrag = '';

                        itemTD.text = itemTD.text || '';


                        if (typeof itemTD.text == 'function') {
                              var funContent = itemTD.text.toString().match(/\/\*([\s\S]+)\*\//)[1], convertedContent;

                              //处理空格缩进的问题

                              funContent =  funContent.replace(/\n{2,}/mg, '\n');

                              var matchSpaceArr = funContent.match(/\n\s+/mg);
                              var spaceCountObj = {}, spaceCountArr = [];
                              $.each(matchSpaceArr, function(_, s) {
                                    //if(spaceCountArr.indexOf(s.length) == -1) {
                                    if($.inArray(s.length, spaceCountArr) == -1) {

                                        spaceCountArr.push(s.length);
                                    }
                                    spaceCountArr.sort();
                                    spaceCountObj[s.length] = '';
                              });

                              var spaceTabIndex = '  ';
                              $.each(spaceCountArr, function(_, spacelen) {
                                    spaceCountObj[spacelen] = spaceTabIndex;
                                    spaceTabIndex += '  ';
                              });

                              convertedContent = funContent.replace(/(\n\s{2,})/mg, function (_, space) {
                                    return '\n' + spaceCountObj[space.length];
                              });
                              console.log(convertedContent);
                              itemTD.text = convertedContent;
                        }

                        if (itemTD.className) {
                              classFrag = ' class="' + itemTD.className + '" ';
                        }

                        if (itemTD.style) {
                              styleFrag = ' style="';
                              for (var cssKey in itemTD.style) {
                                    styleFrag += cssKey + ':' + itemTD.style[cssKey] + ';';
                              }
                              styleFrag += ' "';
                        }
                        if (itemTD.attrs) {
                              for (var attr in itemTD.attrs) {
                                    attrsFrag += ' ' + attr + '=' + "\"" + itemTD.attrs[attr] + "\" ";
                              }
                        }

                        if (itemTD.strip === false) {
                              var textConvert = itemTD.text;
                        }
                        else {
                              var textConvert = $('<div>').text(itemTD.text).html();
                              console.log(textConvert)
                        }


                        if (itemTD.strong) {
                              if ($.type(itemTD.strong) == 'array') {
                                    $.each(itemTD.strong, function (_, s) {

                                          if (itemTD.strongReg === false) {
                                                var sReg = s;
                                                textConvert = textConvert.replace(sReg, '<span class="ft-orange">' + s + '</span>');
                                          }
                                          else {
                                                if (Object.prototype.toString.call(s) == '[object Array]') {
                                                      var sReg = new RegExp(s[0], 'igm');
                                                      textConvert = textConvert.replace(sReg, '<span class="ft-orange">' + s[1] + '</span>');
                                                }
                                                else if (typeof s == 'string') {
                                                      var sReg = new RegExp(s, 'igm');
                                                      textConvert = textConvert.replace(sReg, '<span class="ft-orange">' + s + '</span>');
                                                }
                                                else {
                                                      throw new Error('Strong type Error');
                                                }
                                          }
                                    })
                              }
                              else if (typeof itemTD.strong == 'string') {
                                    textConvert = textConvert.replace(itemTD.strong, '<span class="ft-orange">' + itemTD.strong + '</span>');
                              }
                        }


                        if (itemTD.code === true) {
                              textConvert = '<pre>' + textConvert + '</pre>';
                        }

                        tr += '<td ' + classFrag + styleFrag + attrsFrag + '>' + textConvert + '</td>';
                  });
                  tr += '</tr>';

                  body += tr;
            });

            body += '</tbody>';
            return body;
      }

      function renderTable(head, body) {
            return '<table>' + head + body + '</table>';
      }

}

widget/mask.js
define('src/mask', function(require, exports, module){
      var $ = require("jquery"),
            Overlay = require("./overlay.js"),


            ua = (window.navigator.userAgent || "").toLowerCase(),
            isIE6 = ua.indexOf("msie 6") !== -1,


            body = $(document.body),
            doc = $(document);


// Mask
// ----------
// 全屏遮罩层组件
      var Mask = Overlay.extend({

            attrs: {
                  width: isIE6 ? doc.outerWidth(true) : '100%',
                  height: isIE6 ? doc.outerHeight(true) : '100%',

                  className: 'ui-mask',
                  opacity: 0.2,
                  backgroundColor: '#000',
                  style: {
                        position: isIE6 ? 'absolute' : 'fixed',
                        top: 0,
                        left: 0
                  },

                  align: {
                        // undefined 表示相对于当前可视范围定位
                        baseElement: isIE6 ? body : undefined
                  }
            },

            show: function () {
                  if (isIE6) {
                        this.set('width', doc.outerWidth(true));
                        this.set('height', doc.outerHeight(true));
                  }
                  return Mask.superclass.show.call(this);
            },

            _onRenderBackgroundColor: function (val) {
                  this.element.css('backgroundColor', val);
            },

            _onRenderOpacity: function (val) {
                  this.element.css('opacity', val);
            }
      });

// 单例
      module.exports = new Mask();
});


widget/Dialog.js
/*
 * @author: human huang
 * @email:  halfthink@163.com
 * @created date: 2015-04-08
 */

var Dialog = (function ($, window, undefined) {
      var DiaLogID = 1;

      var InnerUse = function () {
      };

      var Panel = createClass(InnerUse);


      //Action
      Panel.defaults = {
            trigger: null,
            //element: null,

            //Action
            hasMask: true,
            hasEffect: true,
            fixed: true,
            isLock: false,
            quickClose: true,
            timeClose: 0,

            _timeCloseTimer: null,
            _isShow: false,
            //Callback
            onClose: function () {
            }
      };

      //UI
      $.extend(Panel.defaults, {
            title: null,
            content: null,

            width: 500,
            height: 200,
            classPrefix: 'kind-',
            DialogTPL: [
                  '<div class="{classPrefix}dialog {classPrefix}dialog-'+ DiaLogID +'">',
                  '<a class="{classPrefix}dialog-close">{closeTPL}</a>',
                  '<div class="{classPrefix}dialog-container">',
                  '<div class="{classPrefix}dialog-title">{title}</div>',
                  '<div class="{classPrefix}dialog-content">{content}</div>',
                  '</div>',
                  '</div>'
            ].join(''),
            closeTPL: '×',
            zIndex:999
      });

      // Output Interface
      $.extend(Panel.prototype, {

            hide: function () {
                  this.hidePanel();
            },
            show: function () {
                  this.showPanel();
            },
            set: function (attr, value) {
                  if (typeof attr == 'object') {
                        for (var i in attr) {
                              this[i] = attr[i];
                        }
                  }
                  else {
                        this[attr] = value;

                  }
                  return this;
            },
            resize: function() {
                  console.log('resize');
                  var winRect = this.getWinRect();
                  this.box.css({
                     left:winRect.width / 2- this.width / 2,
                     top:winRect.height / 2- this.height / 2
                  });
            },
            destroy: function() {
                  //$( this.trigger ).off();
            }

      });

      //init
      $.extend(InnerUse.prototype, {
            _init: function (opt) {

                  $.extend(this, Panel.defaults, opt);

                  this.parseElement();
                  this.bindEvent();
                  DiaLogID++;

                  this.trigger[0].kindDialog = this;
            },
            parseElement: function () {
                  this.trigger = $(this.trigger);
            }
      });


      //event
      $.extend(InnerUse.prototype, {
            bindEvent: function () {
                  //delay delegate event
                  this.trigger.click($.proxy(this.triggerClick, this));
            },
            triggerClick: function (e) {

                  if (this.trigger.data('kind-dialog-created') == 'true') {
                        this.showPanel();
                        return;
                  }

                  this.renderPanel();

                  //showPanel的时候处理事件
                  this.showPanel();

            },
            onCloseClick: function (e) {
                  this.hidePanel();
            },
            onDocClick: function (e) {
                  console.log('onDocClick');
                  if(this._isShow == false) {
                        return ;
                  }

                  var $target = $(e.target),
                        className = $target[0].className,
                        nodeName = $target[0].nodeName;

                  if (className.indexOf('kind-dialog-mask') != -1 || nodeName == 'HTML' || nodeName == 'BODY') {
                        // mask
                        this.hidePanel();
                  }
                  else if (className.indexOf('kind-dialog') != -1) {

                  }
            },
            onDocKeyup: function (e) {
                  //esc
                  if (e.keyCode == 27) {
                        this.hidePanel();
                  }
            }
      });

      //UI
      $.extend(InnerUse.prototype, {
            addBox: function () {

                  var docRect = this.getWinRect();

                  var injectObj = {
                        title: this.title,
                        content: this.content,
                        closeTPL: this.closeTPL,
                        classPrefix:this.classPrefix
                  };

                  var renderedDialogFrag = this.DialogTPL.replace(/{(\w+)}/mg, function (_, word) {
                        return injectObj[word];
                  });

                  this.box = $(renderedDialogFrag)
                        .css({
                              position: 'fixed',
                              top: docRect.height / 2 - this.height / 2,
                              left: docRect.width / 2 - this.width / 2,
                              width: this.width,
                              height: this.height,
                              zIndex:this.zIndex + 1
                        })
                        .hide().appendTo(document.body);

                  //no Title
                  if( !this.title) {
                        this.box.find('.' + this.classPrefix + 'dialog-title').remove();
                  }

                  return this;
            },

            addMask: function () {
                  var $maskOrigin = $('.' + this.classPrefix + 'dialog-mask');
                  if($maskOrigin[0]) {
                        this.mask = $maskOrigin;
                        return this;
                  }

                  this.mask = $('<div>')
                        .addClass(this.classPrefix + 'dialog-mask')
                        .css({
                              position: 'fixed',
                              top: 0,
                              left: 0,
                              width: '100%',
                              height: '100%',
                              background: '#000',
                              opacity: '0.2',
                              zIndex: this.zIndex
                        })
                        .hide()
                        .appendTo('body');
                  return this;
            },

            renderPanel: function (e) {
                  this.addMask();
                  this.addBox();
                  this.trigger.data('kind-dialog-created', 'true');
            },

            showPanel: function () {

                  //timer close
                  if (this.timeClose) {
                        this._timeCloseTimer = setTimeout($.proxy(this.hidePanel, this), this.timeClose * 1000);
                  }

                  //close
                  if (this.quickClose) {
                        $(document).on('click.kDialog', $.proxy(this.onDocClick, this));
                  }
                  //close
                  $(document).on('keyup.kDialog', $.proxy(this.onDocKeyup, this));

                  //close
                  this.closeButton = $(this.box).find('.kind-dialog-close').on('click.kDialog', $.proxy(this.onCloseClick, this));

                  //resize
                  $(window).on('resize.kDialog', $.proxy(this.resize, this));

                  // lock scroll
                  if (this.isLock) {
                        $(window).on('mousewheel.kDialog DOMMouseScroll.kDialog', function (e) {
                              e.preventDefault();
                        });
                  }

                  this.showMask();
                  this.showBox();

            },

            hidePanel: function () {
                  var onCloseResult;
                  if (this.onClose) {
                        onCloseResult = this.onClose();
                  }
                  if (onCloseResult === false) {
                        return;
                  }

                  // cancel lock scroll
                  if (this.isLock) {
                        $(window).off('.kDialog');
                  }
                  // off Event
                  $(document).off('.kDialog');
                  $(this.box).off('.kDialog');
                  $(this.closeButton).off('.kDialog');


                  this.hideBox();
                  this.hideMask();
            },


            showMask: function () {
                  if (this.hasMask) {
                        this.mask.show();
                  }
            },
            hideMask: function () {
                  if (this.hasMask) {
                        this.mask.hide();
                  }
            },

            showBox: function () {

                  this._isShow = true;

                  if (this.hasEffect) {
                        this.box.fadeIn('fast');
                  }
                  else {
                        this.box.show();
                  }
            },
            hideBox: function () {
                  this._isShow = false;
                  this.box.hide();
            }

      });


      //util
      $.extend(InnerUse.prototype, {
            getWinRect: function () {
                  return {
                        width: $(window).width(),
                        height: $(window).height()
                  };
            },
            getDocRect: function () {
                  return {
                        width: $(document).width(),
                        height: $(document).height()
                  };
            }
      });

      function createClass(Father) {
            function F() {
            }

            function Sub() {
                  if (this.superClass['_init']) {
                        this.superClass._init.apply(this, arguments);
                  }

                  if (this['_init'] && this.hasOwnProperty('_init')) {
                        this._init.apply(this, arguments);
                  }
            }

            F.prototype = Father.prototype;
            Sub.prototype = new F;
            Sub.prototype.constructor = Sub;
            Sub.prototype.superClass = Father.prototype;
            return Sub;
      }

      return Panel;

})(jQuery, window);